File: NOIP2010/图论/最短路径/dijkstra堆优化/dijkstra.cpp
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <algorithm>
using namespace std;
const int maxn=1000;
int n,tot,i,j,now,x[maxn][maxn],y[maxn][maxn],heap[maxn],pos[maxn],key[maxn],sum[maxn];
bool exist[maxn];
void modify(int id,int newkey)
{
    int p;
    if (newkey<key[id])
    {
        key[id]=newkey;
        p=pos[id];
        while (p>1 && key[heap[p]]<key[heap[p/2]]) 
        {
            swap(pos[heap[p]],pos[heap[p/2]]);
            swap(heap[p],heap[p/2]);
            p=p/2;
        }    
    }    
}    
void heapify(int p)
{
    int best;
    best=p;
    while (p*2<=tot && key[heap[p*2]]<key[heap[best]]) best=p*2;
    while (p*2+1<=tot && key[heap[p*2+1]]<key[heap[best]]) best=p*2+1;
    if (best!=p)
    {
        swap(pos[heap[p]],pos[heap[best]]);
        swap(heap[p],heap[best]);
        heapify(best);
    }    
}    
void extract()
{
    int id,dis;
    id=heap[1];
    dis=key[id];
    i=id; now=dis;
    tot--;
    if (tot>0)
    {
        swap(pos[heap[1]],pos[heap[tot+1]]);
        swap(heap[1],heap[tot+1]);
        heapify(1);
    }    
}    
int main()
{
    freopen("dijkstra.in","r",stdin);
    freopen("dijkstra.out","w",stdout);
    scanf("%d",&n);
    for (i=1;i<=n;i++)
    {
        scanf("%d",&sum[i]);
        for (j=1;j<=sum[i];j++) scanf("%d %d",&x[i][j],&y[i][j]);    
    }    
    tot=n; 
    for (i=1;i<=n;i++)
    {
        pos[i]=i;
        heap[i]=i;
        exist[i]=true;
        key[i]=32768;
    }    
    key[1]=0;
    while (tot>0)
    {
        extract();
        if (now==32768)break;
        for (j=1;j<=sum[i];j++)
         modify(x[i][j],now+y[i][j]);
    }    
    printf("%d",key[n]);
    return 0;
}    

File: NOIP2010/图论/最短路径/dijkstra堆优化/最优版本/dijkstra.pas
var now,t,tot,n,m,s,v,x,y,w,cnt,i,j:longint;
    exist:array[1..100]of boolean;
    pos,key,heap:array[1..100]of longint;
    link,value:array[1..100,0..100]of longint;
procedure modify(id,newkey:longint);
var p,tmp:longint;
begin
if newkey<key[id] then
   begin
   key[id]:=newkey;
   p:=pos[id];
   while (p>1)and(key[heap[p]]<key[heap[p div 2]])do
     begin
     tmp:=pos[heap[p]];pos[heap[p]]:=pos[heap[p div 2]];pos[heap[p div 2]]:=tmp;
     tmp:=heap[p]; heap[p]:=heap[p div 2];heap[p div 2]:=tmp;
     p:=p div 2;
     end;
   end;
end;
procedure heapify(q:longint);
var t1,t2,ox,tmp,p:longint;
begin
t1:=maxlongint; t2:=maxlongint;
p:=q;
while (p*2<=tot) do
  begin
  t1:=key[heap[p*2]]; if (p*2+1<=tot) then t2:=key[heap[p*2+1]];
  if t1<t2 then ox:=0 else ox:=1;
  tmp:=heap[p];heap[p]:=heap[p*2+ox];heap[p*2+ox]:=tmp;
  tmp:=pos[heap[p]];pos[heap[p]]:=pos[heap[p*2+ox]];pos[heap[p*2+ox]]:=tmp;
  p:=p*2;
  end;
end;
procedure extract;
begin
now:=key[heap[1]];
t:=heap[1];
heap[1]:=heap[tot]; pos[heap[1]]:=pos[heap[tot]];
exist[t]:=false;
dec(tot);
if tot>0 then heapify(1);
end;
begin
assign(input,'dijkstra.in');
reset(input);
assign(output,'dijkstra.out');
rewrite(output);
readln(n,s,v);
readln(m);
for i:=1 to m do
  begin
  readln(x,y,w);
  cnt:=link[x,0]+1;
  link[x,cnt]:=y;
  inc(link[x,0]);
  value[x,y]:=w;
  end;
for i:=1 to n do
    begin
    key[i]:=maxlongint;
    pos[i]:=i;
    heap[i]:=i;
    exist[i]:=true;
    end;
tot:=n;
key[s]:=0;
while (tot>0) do
  begin
  extract;
  if now=maxlongint then break;
  for i:=1 to link[t,0] do
      begin
      if exist[link[t,i]]=true then modify(link[t,i],now+value[t,link[t,i]]);
      end;
  end;
writeln(key[v]);
close(input);
close(output);
end.